### PR TITLE
HIVE-27987: Implement YESTERDAY UDF in Hive

### PR DESCRIPTION
**Pull request by MRKKmrkk**
### What changes were proposed in this pull request?
Implement YESTERDAY UDF in Hive

### Why are the changes needed?
In my daily work, I frequently need to examine data from the previous day. Currently, in Hive, the only available method to obtain the date of yesterday is by using calculations like date_sub(current_date, 1). It would be convenient if Hive could introduce a method named 'yesterday' to facilitate the quick and easy retrieval of the date corresponding to the previous day.

### Does this PR introduce _any_ user-facing change?
No

### Is the change a dependency upgrade?
No

### How was this patch tested?
Created  qtests as part of this change

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    "The function YESTERDAY does not take any arguments, but found "
+                            + arguments.length);
+        }
+
+        if (yesterday == null) {
+            SessionState ss = SessionState.get();
+            ZonedDateTime dateTime = ss.getQueryCurrentTimestamp().atZone(
+                    ss.getConf().getLocalTimeZone()).minusDays(1L);
+            Date dateVal = Date.valueOf(
+                    dateTime.toString().substring(0, 10));
```
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java:50]: I think it is more straightforward and robust to use this method in this UDF and CURRENT_DATE.
https://github.com/apache/hive/blob/rel/release-4.0.0-beta-1/common/src/java/org/apache/hadoop/hive/common/type/Date.java#L205-L207

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
+package org.apache.hadoop.hive.ql.udf.generic;
```
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java:1]: We need a license.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java [hunk 3]
```diff
@@ -452,6 +452,7 @@ public final class FunctionRegistry {
     system.registerGenericUDF("current_schema", GenericUDFCurrentSchema.class);
     system.registerGenericUDF("current_catalog", GenericUDFCurrentCatalog.class);
     system.registerGenericUDF("current_date", GenericUDFCurrentDate.class);
+    system.registerGenericUDF("yesterday", GenericUDFYesterday.class);
```
Comment (okumin) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java:455]: We need to update this file when we add a new UDF.
https://github.com/okumin/hive/commits/master/ql/src/test/results/clientpositive/llap/show_functions.q.out

**Comment by okumin** [comment]
Googling some database systems, I didn't find ones with YESTERDAY or a similar UDF. But I imagine this UDF would be helpful

**Comment by MRKKmrkk** [comment]
> Googling some database systems, I didn't find ones with YESTERDAY or a similar UDF. But I imagine this UDF would be helpful at this point. I'm not confident though.

**Comment by MRKKmrkk** [comment]
> Googling some database systems, I didn't find ones with YESTERDAY or a similar UDF. But I imagine this UDF would be helpful at this point. I'm not confident though.

Thank you for your review! I understand your confusion, actually clickhouse has similar UDF: https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#yesterday, and this function is widely used by both me and my colleagues in our daily work.

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java (hunk 1 above)
Comment (MRKKmrkk) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java]: Done!

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java (hunk 2 above)
Comment (MRKKmrkk) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFYesterday.java:18]: Done!

File: ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java (hunk 3 above)
Comment (MRKKmrkk) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/FunctionRegistry.java:455]: Done!

**Comment by MRKKmrkk** [comment]
@okumin Could you please check if this fix is ok? Thanks.

**Comment by okumin** [comment]
@MRKKmrkk Thanks for sharing the use case. I'm putting some thoughts here.

1. CI is failing. Could you please check [this](http://ci.hive.apache.org/blue/rest/organizations/jenkins/pipelines/hive-precommit/branches/PR-4985/runs/4/nodes/520/log/?start=0) and [this](http://ci.hive.apache.org/blue/rest/organizations/jenkins/pipelines/hive-precommit/branches/PR-4985/runs/4/nodes/567/log/?start=0)? I guess you need to generate the result using `TestMiniLlapLocalCliDriver`
2. I put some minor comments
3. I can personally give +1 if we resolve the above two points. But note that I am not a committer and you would need to find another person to take another look

**Pull request closed without merging**